doc: nrf: add documentation for PSM usage (#10300)
nrfconnect/sdk-nrf: doc/nrf/app_dev/optimizing/power.rst

An nRF9160 DK can draw current ranging from a few micro amperes (in sleep mode) to hundreds of milli amperes (when the radio is active).
To achieve long battery life, it is crucial that the application is optimized in the use of the radio.

PSM usage
=========

Power Saving Mode (PSM) enables the modem to enter deep sleep when it does not need to communicate with the network.
The modem can spend most of the time in deep sleep, only waking up when data needs to be sent and received.

> When implementing an application taking advantage of the PSM, keep in mind that the enabling PSM request does not mean that modem can always enter PSM.

Reviewer: ```suggestion
When implementing an application that uses PSM, keep in mind that the enabling PSM request does not mean that the modem can always enter PSM.
```

Author: Discussed offline with @trantanen that I'll change this to "requesting PSM".

> Possible cases when PSM can not be used:

Reviewer: ```suggestion
Possible cases when PSM cannot be used:
```

* Network does not support PSM.
* Roaming network supports PSM, but does not allow the roaming subscription to use it.

> * Modem does not find a suitable cell.

Author: Neither. The term "suitable cell" comes from 3GPP specifications and means a cell which provides normal service for the device.

https://www.3glteinfo.com/suitable-cell-in-3gpp-umts-system/

> * Modem does not find a suitable cell.

Reviewer: Sounds ok to me.

You can detect the first two cases by checking the ``+CEREG`` notification or ``AT%XMONITOR`` command output for Active-Time granted by the network.

> If the Active-Time is ``11100000``, the timer is deactivated and PSM can not be used.
> Depending on the use case, it may be preferable to set modem to offline mode during long periods of inactivity to conserve power when PSM is not available.

Reviewer: ```suggestion
If the Active-Time is ``11100000``, the timer is deactivated and PSM cannot be used.
Depending on the use case, it may be preferable to set the modem to offline mode during long periods of inactivity to save power when PSM is not available.
```

The third case can happen when cellular network coverage is poor.

> After modem has entered PSM, it stays in deep sleep regardless of cellular network availability.
> When modem wakes up from PSM to transmit data or to do a periodic Tracking Area Update (TAU), it starts searching for a suitable cell.
> If a suitable cell is not found during the initial search, modem will continue periodic cell searches until a suitable cell is found.
> Because modem can only enter PSM when it’s in registered state and camped on suitable cell, the modem can not enter PSM until a suitable cell has been found.

Reviewer: ```suggestion
After the modem has entered PSM, it stays in deep sleep regardless of cellular network availability.
When the modem wakes up from PSM to transmit data or to do a periodic Tracking Area Update (TAU), it starts searching for a suitable cell.
If a suitable cell is not found during the initial search, the modem will continue periodic cell searches until a suitable cell is found.
Because modem can only enter PSM when it’s in registered state and camped on suitable cell, the modem can not enter PSM until a suitable cell has been found.
```

> There are a couple of ways to mitigate the increased power consumption caused by poor cellular coverage.

Reviewer: ```suggestion
The following two sections introduce ways to mitigate the increased power consumption caused by poor cellular coverage.
```

Author: @peknis I left the word "two" out for the reason suggested by @trantanen above.

Making periodic cell searches less frequent
-------------------------------------------

You can modify the periodic cell search configuration to perform periodic searches less frequently.

> You can do this by using the ``AT%XDATAPRFL`` command to set the data profile to "ultra-low power" or modifying the cell search configuration directly using ``AT%PERIODICSEARCHCONF``.

Reviewer: ```suggestion
Use the ``AT%XDATAPRFL`` command to set the data profile to "ultra-low power" or modify the cell search configuration directly using ``AT%PERIODICSEARCHCONF``.
```

For example, you can configure periodic searches to be performed infrequently and when data connectivity is needed, trigger an extra search immediately using ``AT%PERIODICSEARCHCONF=3``.

> In addition to periodic cell searches, the modem may remain in limited service on a cell which doesn't provide normal service for the device.
> This prevents the modem from going to deep sleep between periodic cell searches.

Reviewer: ```suggestion

In addition to periodic cell searches, the modem may remain in limited service on a cell that does not provide normal service for the device.
This prevents the modem from going to deep sleep between periodic cell searches.
```

> Setting modem to offline mode when it can not enter PSM
> -------------------------------------------------------

Reviewer: ```suggestion
Setting modem to offline mode when it cannot enter PSM
------------------------------------------------------
```

> If modem does not enter PSM as expected, the application can set the modem to the offline mode to prevent excessive current consumption when modem can not find a suitable cell.
> You can detect this by monitoring the modem sleep notifications.

Reviewer: ```suggestion
If the modem does not enter PSM as expected, the application can set the modem to offline mode to prevent excessive current consumption when it cannot find a suitable cell.
To detect this, monitor the modem sleep notifications.
```

Preconditions:

* PSM has been negotiated with the network.
* ``+CSCON`` RRC connection status notifications have been subscribed.
* ``%XMODEMSLEEP`` modem sleep notifications have been subscribed.
* User data communication has been completed.

> Steps to check whether modem is able to activate PSM after RRC connection release:

Reviewer: ```suggestion
To check whether the modem is able to activate PSM after RRC connection release, complete the following steps:
```

1. Wait for AT notification indicating that RRC connection was released (``+CSCON: 0``).
#. Read PSM active time using ``AT%XMONITOR``.

>    The active time is used in step 3.

Reviewer: ```suggestion
   The active time is used in Step 3.
```

#. Wait for ``%XMODEMSLEEP: 1`` notification indicating that modem has entered PSM.

>    If this notification is not received within PSM active time (allow for some margin), modem was not able to enter PSM.

Reviewer: ```suggestion
   If this notification is not received within PSM active time (allow for some margin), the modem was not able to enter PSM.
```

>    If ``%XMODEMSLEEP: 3`` notification is received while waiting for PSM, the cell has been lost and modem can not enter PSM.
>    You can use this as a trigger to set modem to offline mode quicker in this case, but it's not mandatory.

Reviewer: ```suggestion
   If ``%XMODEMSLEEP: 3`` notification is received while waiting for PSM, the cell has been lost and the modem cannot enter PSM.
   You can use this as a trigger to set the modem to offline mode quicker in this case, but it is not mandatory.
```

> #. If RRC connection is activated again (``+CSCON: 1``) while waiting for PSM, go back to step 1.

Reviewer: ```suggestion
#. If RRC connection is activated again (``+CSCON: 1``) while waiting for PSM, go back to Step 1.
```
